include/awkward/Slice.h

```diff
@@ -1,179 +1,198 @@
-    }
-    const std::string tostring_part() const;
-    const IndexOf<T> ravel() const;
-  private:
-    const IndexOf<T> index_;
-    const std::vector<int64_t> shape_;
-    const std::vector<int64_t> strides_;
-  };
-
-  typedef SliceArrayOf<int64_t> SliceArray64;
-
-  class SliceField: public SliceItem {
-  public:
-    SliceField(const std::string& key): key_(key) { }
-    const std::string key() const { return key_; }
-    virtual const std::shared_ptr<SliceItem> shallow_copy() const {
-      return std::shared_ptr<SliceItem>(new SliceField(key_));
-    }
-    virtual const std::string tostring() const;
-    virtual bool preserves_type(const std::shared_ptr<Type>& type, const Index64& advanced) const {
-      return false;
-    }
-  private:
-    const std::string key_;
-  };
-
-  class SliceFields: public SliceItem {
-  public:
-    SliceFields(const std::vector<std::string>& keys): keys_(keys) { }
-    const std::vector<std::string> keys() const { return keys_; }
-    virtual const std::shared_ptr<SliceItem> shallow_copy() const {
-      return std::shared_ptr<SliceItem>(new SliceFields(keys_));
-    }
-    virtual const std::string tostring() const;
-    virtual bool preserves_type(const std::shared_ptr<Type>& type, const Index64& advanced) const {
-      return type.get() != nullptr  &&  type.get()->numfields() != -1  &&  util::subset(keys_, type.get()->keys());
-    }
-  private:
-    const std::vector<std::string> keys_;
-  };
-
-  class Slice {
-  public:
-    static int64_t none() { return SliceItem::none(); }
-
-    Slice(): items_(std::vector<std::shared_ptr<SliceItem>>()), sealed_(false) { }
-    Slice(const std::vector<std::shared_ptr<SliceItem>> items): items_(items), sealed_(false) { }
-    Slice(const std::vector<std::shared_ptr<SliceItem>> items, bool sealed): items_(items), sealed_(sealed) { }
-    const std::vector<std::shared_ptr<SliceItem>> items() const { return items_; }
-    bool sealed() const { return sealed_; }
-
-    int64_t length() const;
-    int64_t dimlength() const;
-    const std::shared_ptr<SliceItem> head() const;
-    const Slice tail() const;
-    const std::string tostring() const;
-    void append(const std::shared_ptr<SliceItem>& item);
-    void append(const SliceAt& item);
-    void append(const SliceRange& item);
-    void append(const SliceEllipsis& item);
-    void append(const SliceNewAxis& item);
-    template <typename T>
-    void append(const SliceArrayOf<T>& item);
-    void become_sealed();
-    bool isadvanced() const;
-
-  private:
-    std::vector<std::shared_ptr<SliceItem>> items_;
-    bool sealed_;
-  };
-}
+class SliceItem {
+public:
+  static int64_t none() { return kSliceNone; }
+
+  virtual ~SliceItem() {}
+
+  virtual const std::shared_ptr<SliceItem> shallow_copy() const = 0;
+  virtual const std::string tostring() const = 0;
+  virtual bool preserves_type(const std::shared_ptr<Type> &type,
+                              const Index64 &advanced) const = 0;
```

Comment:
This kind of linewidth-sculpting makes it less clear. Maybe we just want a larger linewidth. GitHub has a natural linewidth and we're often viewing this code in GitHub, so maybe that should be the linewidth.

(Incidentally, I determined the maximum width of the string-view of `awkward1.Array` objects (in Python) to be such that it fits into GitHub and StackOverflow text boxes without scrolling. That's what's really important!)